record/record_test.go

```diff
@@ -636,3 +642,138 @@ func TestRecoverLastCompleteBlock(t *testing.T) {
+		// The first record will consume 3 entire blocks but a fraction of the 4th.
+		blockSize*3,
+		// The second record will completely fill the remainder of the 4th block.
+		3*(blockSize-headerSize)-2*blockSize-2*headerSize,
+		// Consume the entirety of the 5th block.
+		blockSize-headerSize,
+		// Consume the entirety of the 6th block.
+		blockSize-headerSize,
+		// Consume roughly half of the 7th block.
+		blockSize/2,
+	)
+	if err != nil {
+		t.Fatalf("makeTestRecords: %v", err)
+	}
+
+	r := NewReader(bytes.NewReader(recs.buf))
+	// Seek to a valid block offset, but within a multiblock record. This should cause the next call to
+	// Next after SeekRecord to return the next valid FIRST/FULL chunk of the subsequent record.
+	err = r.SeekRecord(blockSize)
+	if err != nil {
+		t.Fatalf("SeekRecord: %v", err)
+	}
+	rec, err := r.Next()
+	if err != nil {
+		t.Fatalf("Next: %v", err)
+	}
+	rData, _ := ioutil.ReadAll(rec)
+	if !bytes.Equal(rData, recs.records[1]) {
+		t.Fatalf("Unexpected output in record 1's data, got %v want %v", rData, recs.records[1])
+	}
+
+	// Seek to 3 bytes into the second block, which is still in the middle of the first record, but not
+	// at a valid chunk boundary.
+	err = r.SeekRecord(blockSize + 3)
+	if err != nil {
+		t.Fatalf("SeekRecord: %v", err)
+	}
+
+	if _, err = r.Next(); err == nil {
+		t.Fatalf("Unexpected lack of error calling SeekRecord to a bad offset")
+	}
+	r.Recover() // Recover from the faux error just generated.
+
+	// Seek to the fifth block and verify all records can be read as appropriate.
+	err = r.SeekRecord(blockSize * 4)
+	if err != nil {
+		t.Fatalf("SeekRecord: %v", err)
+	}
+
+	for i := 2; i < len(recs.records); i++ {
+		rec, err := r.Next()
+		if err != nil {
+			t.Fatalf("Next: %v", err)
+		}
+
+		rData, _ := ioutil.ReadAll(rec)
+		if !bytes.Equal(rData, recs.records[i]) {
+			t.Fatalf("Unexpected output in record #%d's data, got %v want %v", i, rData, recs.records[i])
+		}
+	}
+
+	// Seek back to the fourth block, and read all subsequent records and verify them.
+	err = r.SeekRecord(blockSize * 3)
+	if err != nil {
+		t.Fatalf("SeekRecord: %v", err)
+	}
+
+	for i := 1; i < len(recs.records); i++ {
+		rec, err := r.Next()
+		if err != nil {
+			t.Fatalf("Next: %v", err)
+		}
+
+		rData, _ := ioutil.ReadAll(rec)
+		if !bytes.Equal(rData, recs.records[i]) {
+			t.Fatalf("Unexpected output in record #%d's data, got %v want %v", i, rData, recs.records[i])
+		}
+	}
+
+	// Now seek past the end of the file and verify it causes an error.
+	err = r.SeekRecord(1 << 20)
+	if err == nil {
+		t.Fatalf("Seek past the end of a file didn't cause an error")
+	}
+	if err != io.EOF {
+		t.Fatalf("Seeking past EOF raised unexpected error: %v", err)
+	}
+	r.Recover() // Verify recovery works.
+
+	// Validate the current records are returned after seeking to a valid offset.
+	err = r.SeekRecord(blockSize * 4)
+	if err != nil {
+		t.Fatalf("SeekRecord: %v", err)
+	}
+
+	for i := 2; i < len(recs.records); i++ {
```

Comment:
You've copy/pasted this loop three times. It seems like you could make a closure and just call it three times.

It doesn't have to be a top-level function, and being a closure, you can close over instead of explicitly passing t and r as arguments.